modules/alloy/src/main/ts/ephox/alloy/positioning/layout/Layout.ts

```diff
@@ -47,36 +52,91 @@ const westEdgeX = (anchor: AnchorBox, element: AnchorElement): number => {
   return anchor.x() - element.width();
 };
 
-const southeast: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble) => {
-  return NuSpotInfo(eastX(anchor), southY(anchor), bubbles.southeast(), Direction.southeast(), 'layout-se');
-};
-
-const southwest: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble) => {
-  return NuSpotInfo(westX(anchor, element), southY(anchor), bubbles.southwest(), Direction.southwest(), 'layout-sw');
-};
-
-const northeast: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble) => {
-  return NuSpotInfo(eastX(anchor), northY(anchor, element), bubbles.northeast(), Direction.northeast(), 'layout-ne');
-};
-
-const northwest: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble) => {
-  return NuSpotInfo(westX(anchor, element), northY(anchor, element), bubbles.northwest(), Direction.northwest(), 'layout-nw');
-};
-
-const north: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble) => {
-  return NuSpotInfo(middleX(anchor, element), northY(anchor, element), bubbles.north(), Direction.north(), 'layout-n');
-};
-
-const south: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble) => {
-  return NuSpotInfo(middleX(anchor, element), southY(anchor), bubbles.south(), Direction.south(), 'layout-s');
-};
-
-const east: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble) => {
-  return NuSpotInfo(eastEdgeX(anchor), centreY(anchor, element), bubbles.east(), Direction.east(), 'layout-e');
-};
-
-const west: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble) => {
-  return NuSpotInfo(westEdgeX(anchor, element), centreY(anchor, element), bubbles.west(), Direction.west(), 'layout-w');
+const southeast: AnchorLayout = (anchor: AnchorBox, element: AnchorElement, bubbles: Bubble, bounds: Bounds) => {
+  return NuSpotInfo(
+    eastX(anchor),
+    southY(anchor),
+    bubbles.southeast(),
+    Direction.southeast(),
+    adjustBounds(bounds, anchor, bubbles.southeast(), anchorLeft, anchorBottom, identity, identity),
```

Comment:
Instead of this being an immediately calculated value, you could just return information about what to do `boundAdjustments` which would have a left and a right and a bubble. So it would have the five final argument passed in here, as an object, and the calling code would then call adjustBounds with the anchor and bounds it knowns of, and these bound adjustments. Not sure if it would make it clearer or less clear.

Reply:
I guess the advantage of an approach like that is you know that all bounds adjustments work the same way ... so it's just a matter of how a particular layouts adjusts it. Although, I really should look at LinkedLayouts before suggesting this ...

Reply:
Thanks for the suggestion and I'd agree. I've reworked it so that now you just specify the specific bounds restrictions that need to be applied within the layout and then the bounds is actually adjusted during the bounder attempt. Does that look better to you now?